lib/experimental/html/class-wp-html-walker.php

```diff
@@ -0,0 +1,1270 @@
+	}
+
+	/**
+	 * Parses the next attribute.
+	 *
+	 * @param string $context tag-opener or tag-closer.
+	 * @since 6.1.0
+	 */
+	private function parse_next_attribute( $context = 'tag-opener' ) {
+		// Skip whitespace and slashes
+		$this->parsed_bytes += strspn( $this->html, " \t\f\r\n/", $this->parsed_bytes );
+
+		/*
+		 * Treat the equal sign ("=") as a part of the attribute name if it is the
+		 * first encountered byte:
+		 * https://html.spec.whatwg.org/multipage/parsing.html#before-attribute-name-state
+		 */
+		$name_length = '=' === $this->html[ $this->parsed_bytes ]
+			? 1 + strcspn( $this->html, "=/> \t\f\r\n", $this->parsed_bytes + 1 )
+			: strcspn( $this->html, "=/> \t\f\r\n", $this->parsed_bytes );
+
+		// No attribute, just tag closer.
+		if ( 0 === $name_length ) {
+			return false;
+		}
+
+		$attribute_start     = $this->parsed_bytes;
+		$attribute_name      = substr( $this->html, $attribute_start, $name_length );
+		$this->parsed_bytes += $name_length;
+
+		$this->skip_whitespace();
+
+		$has_value = '=' === $this->html[ $this->parsed_bytes ];
+		if ( $has_value ) {
+			$this->parsed_bytes++;
+			$this->skip_whitespace();
+
+			switch ( $this->html[ $this->parsed_bytes ] ) {
+				case "'":
+				case '"':
+					$quote              = $this->html[ $this->parsed_bytes ];
+					$value_start        = $this->parsed_bytes + 1;
+					$value_length       = strcspn( $this->html, $quote, $value_start );
+					$attribute_end      = $value_start + $value_length + 1;
+					$this->parsed_bytes = $attribute_end;
+					break;
+
+				default:
+					$value_start        = $this->parsed_bytes;
+					$value_length       = strcspn( $this->html, "> \t\f\r\n", $value_start );
+					$attribute_end      = $value_start + $value_length;
+					$this->parsed_bytes = $attribute_end;
+			}
+		} else {
+			$value_start   = $this->parsed_bytes;
+			$value_length  = 0;
+			$attribute_end = $attribute_start + $name_length;
+		}
+
+		if ( 'tag-opener' !== $context ) {
+			return true;
+		}
+
+		// If an attribute is listed many times, only use the first declaration and ignore the rest.
+		if ( ! array_key_exists( $attribute_name, $this->attributes ) ) {
+			$this->attributes[ $attribute_name ] = new WP_HTML_Attribute_Token(
+				$attribute_name,
+				$value_start,
+				$value_length,
+				$attribute_start,
+				$attribute_end,
+				! $has_value
+			);
+		}
+
+		return $this->attributes[ $attribute_name ];
+	}
+
+	/**
+	 * Move the pointer past any immediate successive whitespace.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return void
+	 */
+	private function skip_whitespace() {
+		$this->parsed_bytes += strspn( $this->html, " \t\f\r\n", $this->parsed_bytes );
+	}
+
+	/**
+	 * Asserts that the HTML Walker has not been closed for further lookup or modifications.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @throws WP_HTML_Walker_Exception If the HTML Walker has been closed.
+	 */
+	private function assert_not_closed() {
+		if ( $this->closed ) {
+			throw new WP_HTML_Walker_Exception(
```

Comment:
Can we throw a regular `Exception` like [`BadMethodCallException`](https://www.php.net/manual/en/class.badmethodcallexception.php) with a unique code or run `_doing_it_wrong` function to report an error in logs?

I'm just trying to figure out if there is a simple way to avoid creating a new class that doesn't have any custom logic.

Example:

https://github.com/WordPress/wordpress-develop/blob/37368555a5251e49b748bf91407a24fad6d32e9a/src/wp-includes/class-wp-block-type-registry.php#L54-L90